### PR TITLE
Make .bat and .cmd editors work on windows

### LINKED ISSUE #1316: Vim is not available as external editor on windows
Thanks for `gitui`, it makes my daily life easier=)

There is, however, one thing that seems break my current flow - I can not launch vim from within gitui.

**Describe the bug**
Vim is not available as external editor

**To Reproduce**
1. Navigate to a repo with modified files, run `gitui`
2. Chose a file with modifications
3. Hit 'e' button to edit this file
4. A "popup" appears:
```
failed to launch editor:
"vim": program not found
```
Same bug observed when trying to '^e'dit commit message in external editor.

**Expected behavior**
A `vim` instance with selected file loaded.

**Context (please complete the following information):**
 - OS/Distro + Version: win10 (build 21H2-19044.1889)
 - GitUI Version 0.21.0
 - Running in `WindowsTerminal` and `pwsh-7.2.6`
 - both `vim` and `gitui` are installed via `chocolatey`

**Additional context**
`vim` is present in path:
```
PS1> where.exe vim
C:\Windows\vim.bat
```
```console
git config  -l | rg editor
core.editor=vim
```

### PR DESCRIPTION
**Pull request by pm100**
This Pull Request fixes/closes #1316.

This has several parts
- first it adds code to enable .bat and .cmd commands to be invoked without the .bat or .cmd being given (core.editor='vim' will now run vim.bat)
- it then adds unit tests
- it also reorganized some code used for testing in asyncgit so that it can be used in the main, gitui, product

The actual fix is fairly straightforward.
- if on windows and the configured editor fails to launch then retry
- retry using `cmd /C <editor> <file>`
- if that fails to start or cmd reports an error then report that to the user

The reorg needs an explanation
- open_file_in_editor need a repo. 
- there is test plumbing in asyncgit that provides many repo utilities
- sadly things declared 'test' are not available in other products / subcrates. But there is a solution. see https://github.com/rust-lang/cargo/issues/8379. It involves creating a test-utils feature and making that a dev dependency.
- That now makes that test plumbing in asyncgit/sync/mod.rs no longer hidden under a test umbrella. The exhaustive cliuppy, lints, cargo-sort etc now picks up all sorts of issues there (that really dont matter) so now I had to add in a few things to stop those being picked up. Note that they are still only being compiled when doing tests, its just that clippy 'all-features' doesnt realize that

I also note that there are no other external editor test cases. I will add more once I get this in (I willl look at #1654 and put them in the fix for that). I did add one generic test, but there should be more 

I followed the checklist:
- [ x] I added unittests
- [x ] I ran `make check` without errors
- [x ] I tested the overall application
- [ ] I added an appropriate item to the changelog

### CODE REVIEW

File: asyncgit/src/lib.rs [hunk 1]
```diff
@@ -23,6 +23,7 @@
 // #![deny(clippy::expect_used)]
 //TODO: consider cleaning some up and allow specific places
 #![allow(clippy::significant_drop_tightening)]
+#![allow(clippy::multiple_crate_versions)]
```
Comment (extrawurst) [inline asyncgit/src/lib.rs:26]: this is a red-flag. why do we want that? it bloats the binary. did you investigate what introduces the multiple versions of the same crate?

File: deny.toml [hunk 2]
```diff
@@ -22,5 +22,6 @@ version = "1.0.3"
 multiple-versions = "deny"
 skip-tree = [
     { name = "windows-sys" },
-    { name = "hermit-abi" }
+    { name = "hermit-abi" },
+    { name = "syn" }
```
Comment (extrawurst) [inline deny.toml:26]: see above

**Comment by pm100** [comment]
Multiple versions do not get built into the product, they only happen on test builds

`cargo clippy --workspace --all-features` complains about the code that is in the test feature 'test_utils'. It does not realize that this is only for testing

Same for cargo deny

I tried real hard to untangle the multiple versions even though it only really happens with tests. I could not untangle some of them (people who version  their crate 0.xx are very annoying because 0.45 is not seen as a valid replacement for 0.44 by cargo. Yes windows-sys we are looking at you)

Then I saw that you already had the ignore dups in main.rs and in deny.toml and decided OK its not just me that could not un ravel these

TL;DR - the dups only happen when running tests

**Comment by pm100** [comment]
I had to do a quick commit with that allow taken out because I could not remember what actually failed, I knew it was part of the CI build

**Comment by pm100** [comment]
I just discovered squash. I apologize for my messy prs. I will clean this up
